cli/lib/webpack/service.js

```diff
@@ -0,0 +1,233 @@
+
+		this.forceRebuild = this.buildSettings.platform === 'ios' ? builder.forceCleanBuild : builder.forceRebuild;
+		this.jobIdentifier = this.generateJobIdentifier();
+		this.pluginVersion = process.env.APPCD_WEBPACK_VERSION || 'latest';
+		this.basePath = `/webpack/${this.pluginVersion}`;
+		this.webpackStatusEmitter = new EventEmitter();
+		this.webpackStatus = { state: 'unknown' };
+		this.reporter = new WebpackProgressReporter();
+	}
+
+	loadBuildSettings(builder, cli) {
+		const platform = cli.argv.platform;
+		const buildSettings = {
+			platform,
+			deployType: builder.deployType,
+			target: cli.argv.target,
+			sdk: {
+				path: cli.sdk.path,
+				version: cli.sdk.name,
+				gitHash: cli.sdk.manifest.githash,
+				buildDate: cli.sdk.manifest.timestamp
+			}
+		};
+
+		if (platform === 'ios') {
+			buildSettings.ios = {
+				deviceFamily: builder.deviceFamily
+			};
+		}
+
+		this.buildSettings = buildSettings;
+	}
+
+	async ensureDaemonIsRunning() {
+		try {
+			await this.connectDaemon();
+		} catch (e) {
+			if (e.code === 'ECONNREFUSED') {
+				this.logger.info('Starting Appcelerator Daemon (subsequent builds will be faster) ...');
+				await this.connectDaemon({ startDaemon: true });
+			} else {
+				throw e;
+			}
+		}
+
+		const webUiUrl = `http://${this.client.host}:${this.client.port}${this.basePath}/web`;
+		this.logger.info(`🖥  Web UI ready on ${webUiUrl.cyan}`);
+	}
+
+	connectDaemon(options) {
+		return new Promise((resolve, reject) => {
+			this.client.connect(options)
+				.once('connected', resolve)
+				.once('error', reject);
+		});
+	}
+
+	ensureWebpackIsRunning() {
+		const options = {
+			identifier: this.jobIdentifier,
+			type: this.type,
+			task: 'build', // this.buildSettings.deployType === 'production' ? 'build' : 'serve',
+			watch: this.shouldWatch,
+			project: this.projectData,
+			build: this.buildSettings
+		};
+
+		return new Promise((resolve, reject) => {
+			if (this.forceRebuild) {
+				this.startWebpack(options).then(resolve).catch(reject);
+				return;
+			}
+
+			this.client
+				.request({
+					path: `${this.basePath}/status/${options.identifier}`,
+				})
+				.once('response', response => {
+					this.webpackStatus = response;
+					if (this.webpackStatus.state === 'stopped' || this.webpackStatus.state === 'error') {
+						this.logger.debug('Webpack not running, starting new Webpack build.');
+						this.startWebpack(options).then(resolve).catch(reject);
+					} else {
+						if (!this.validateOptions(response.options, options)) {
+							this.logger.debug('Build options changed, restarting Webpack build.');
+							this.startWebpack(options).then(resolve).catch(reject);
+							return;
+						}
+						if (this.webpackStatus.state === 'ready') {
+							this.builder.tiSymbols = response.tiSymbols;
+						}
+						return resolve();
+					}
+				})
+				.once('error', error => {
+					if (error.status === 404) {
+						this.startWebpack(options).then(resolve).catch(reject);
```

Comment:
<!--
  0 failure: 
  2 warning:  cli/lib/webpack/s..., cli/lib/webpack/s...
  
  
  DangerID: danger-id-Danger;
  File: cli/lib/webpack/service.js;
  Line: 116;
-->

- :warning: cli/lib/webpack/service.js line 116 – Avoid using promises inside of callbacks. (promise/no-promise-in-callback)
- :warning: cli/lib/webpack/service.js line 116 – Avoid using promises inside of callbacks. (promise/no-promise-in-callback)